packages/astro/config.d.ts

```diff
@@ -46,3 +48,22 @@ export function passthroughImageService(): ImageServiceConfig;
  * Return a valid env field to use in this Astro config for `experimental.env.schema`.
  */
 export const envField: EnvField;
+
+/**
+ * Recursively merges a partial configuration object into a valid default.
+ *
+ * Configuration can be merged both before and after being resolved.
+ */
+export function mergeConfig<C extends AstroConfig | AstroInlineConfig>(
+	defaults: C,
+	overrides: DeepPartial<C>,
+): C;
+
+/**
+ * Turn raw config values into normalized values
+ */
+export function validateConfig(
+	userConfig: any,
+	root: string,
+	cmd: string,
+): Promise<AstroConfig>;
```

Comment:
What's the usecase for exporting this? I'm not too comfortable with exposing this yet since the validation doesn't take in account of `--verbose` and `--silent` (or log level in the future) and it always logs to the console when it shouldn't in tests, so it requires some more refactor.

Reply:
It's one of the APIs necessary to implement testing properly. Otherwise, we'd need to re-implement the transformations fone in the schema.

It is mentioned at the end of [the proposal](https://github.com/withastro/roadmap/discussions/1021):
> Similarly, the function `validateConfig` that fills default values and apply transformations on the configuration is also not exported. This is required to resolve a value from the configuration that is not directly set on the test but used for the test tolling, like the `outDir` and `srcDir`.

---

> it always logs to the console when it shouldn't in tests, so it requires some more refactor.

I don't see anywhere in the code where it would log anything. It creates a Zod schema and parses using it. None of the `refine`, `superRefine` and `transform` functions in the schema has any logs, they just add possible issues which will be reported as an error to the caller not as logs.

Are you perhaps thinking of `resolveConfig`?

Reply:
Ah yeah it's `resolveConfig` that's logging the error, not `validateConfig`. But in that case, wouldn't `resolveConfig` be more useful as it mimics what `dev()` and `build()` loads for the astro config? `validateConfig()` only handles the validation part.